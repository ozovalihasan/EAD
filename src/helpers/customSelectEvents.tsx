export const handleMouseUpForSelect = (el: React.MutableRefObject<HTMLSelectElement | null>) => {
  el.current!.classList.toggle("hidden")
}

export const handleMouseLeaveForSelect = (el: React.MutableRefObject<HTMLSelectElement | null>) => {
  el.current!.classList.add("hidden")
}