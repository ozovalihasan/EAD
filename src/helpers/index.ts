export * from './setColorVariants';
export * from './customSelectEvents';
