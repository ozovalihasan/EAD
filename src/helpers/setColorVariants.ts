
import colors from "tailwindcss/colors"

export const availableColors = ["sky", "neutral", "green", "red", "yellow", "pink", "orange", "emerald", "violet"] as const;
export type availableColorsType = typeof availableColors[number];

const setColorVariants = (color?: availableColorsType) => {
  if (color) {
    localStorage.setItem("mainColor", JSON.stringify(color))
  } else {
    
    if (localStorage.getItem("mainColor")) {
      const colorFromLocalStorage = JSON.parse(localStorage.getItem("mainColor")!) as string
      if ( availableColors.includes( colorFromLocalStorage as availableColorsType) ) {
        color = colorFromLocalStorage as availableColorsType
      } else {
        color = availableColors[0]
        localStorage.setItem("mainColor", JSON.stringify(color))
      }
    } else {
      color = availableColors[0]
      localStorage.setItem("mainColor", JSON.stringify(color))
    }
  }

  // eslint-disable-next-line @typescript-eslint/non-nullable-type-assertion-style
  const root = document.querySelector('div#root') as HTMLDivElement;

  const colorPalette = colors[color];

  Object.entries(colorPalette).forEach( ( [num, value] ) => {
    root.style.setProperty(`--first-${num}`, value);  
  });
}

export { setColorVariants }