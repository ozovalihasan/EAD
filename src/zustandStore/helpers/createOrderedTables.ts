import { TablesType } from "@/zustandStore"

export const createOrderedTables = (tables: TablesType) => {
  return Object.keys(tables)
}