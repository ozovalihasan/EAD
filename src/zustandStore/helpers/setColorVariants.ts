
import colors from "tailwindcss/colors"

export const availableColors = ["neutral", "green", "sky", "red", "yellow", "pink", "orange", "emerald", "violet"] as const;
export type availableColorsType = typeof availableColors[number];

const setColorVariants = (color: availableColorsType) => {
  // eslint-disable-next-line @typescript-eslint/non-nullable-type-assertion-style
  const root = document.querySelector('div#root') as HTMLDivElement;

  localStorage.setItem("selected-color", color);
 
  const colorPalette = colors[color];

  Object.entries(colorPalette).forEach( ( [num, value] ) => {
    root.style.setProperty(`--first-${num}`, value);  
  });
}

export { setColorVariants }