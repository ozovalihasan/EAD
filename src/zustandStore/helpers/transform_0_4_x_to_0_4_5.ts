import { State } from "@/zustandStore/store";
import { TableValueType } from "@/zustandStore/tables";

export const transform_0_4_x_to_0_4_5 = (data: State) => {
  if (data.version == "0.4.5") {return data}
  
  data.version = "0.4.5";
  
  Object.values(data.tables).forEach((table: TableValueType) => {
    table.superclassId = ""
  })
  
  return data
}

export default transform_0_4_x_to_0_4_5;