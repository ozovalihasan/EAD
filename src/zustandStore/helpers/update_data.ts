import { State, throughEdgePartial } from "@/zustandStore";
import { TableValueType } from "@/zustandStore/tables";

export const update_data = (data: State) => {

  if ( !(["0.4.5", "0.4.6", "0.4.7"].includes( data.version )) ) {
    Object.values(data.tables).forEach((table: TableValueType) => {
      table.superclassId = ""
    })
  }

  if ( !(["0.4.7"].includes( data.version )) ) {
    
    data.edges = data.edges.map(
      edge => edge.type === throughEdgePartial.type ? edge : {...edge, data: {optional: false}} 
    )

    data.nodes = data.nodes.map(
      (node) => ({
        ...node, style: {width: node.width, height: node.height}
      })
    ) 

  }
  data.version = "0.4.7";
  
  return data;
}

export default update_data;