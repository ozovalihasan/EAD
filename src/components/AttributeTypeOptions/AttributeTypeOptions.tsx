

import { handleMouseLeaveForSelect, handleMouseUpForSelect } from '@/helpers';
import useStore from '@/zustandStore/store';
import { memo, useRef } from 'react';

export const AttributeTypeOptions = memo(({tableId, attributeId}: {tableId: string, attributeId: string} ) =>  {

  const type = useStore((state) => state.tables[tableId].attributes[attributeId].type);
  const onAttributeTypeChange = useStore((state) => state.onAttributeTypeChange);
  const options = [
    'primary_key', 'string', 'text', 'integer', 'float', 'decimal', 'datetime', 'timestamp',
    'time', 'date', 'binary', 'boolean', 'references'
  ]

  const selectEl = useRef<HTMLSelectElement | null>(null);

  return (
    <div 
      className='relative whitespace-nowrap w-full'
      onMouseLeave={() => handleMouseLeaveForSelect(selectEl)}
    >
      <div 
        className='truncate p-2 btn-select rounded-md min-w-full'   
        onMouseUp={() => handleMouseUpForSelect(selectEl)} 
      >
        {type}
      </div>
      <select
        ref={selectEl}
        className="hidden absolute right-0 top-full z-10 min-w-full custom-select-options"
        value={type}
        onChange={(event) => onAttributeTypeChange(event, tableId, attributeId)}
        title="Select attribute type"
        size={options.length}
      >

      {
        options.map((item) => (
          <option
            className="p-2 w-full"
            key={item}
            value={item}
            title={item}
          >
            {item}
          </option>
        ))
      }
      </select>
    </div>
  )
})
