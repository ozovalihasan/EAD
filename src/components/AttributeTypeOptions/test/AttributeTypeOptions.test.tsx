import {  AttributeTypeOptions } from '../AttributeTypeOptions';
import { fireEvent, render, renderHook, screen } from "@testing-library/react";
import useStore from '@/zustandStore/store';
import { handleMouseLeaveForSelect, handleMouseUpForSelect } from '@/helpers';

let renderReadyComponent: JSX.Element;

jest.mock('@/helpers',  () => ({
  handleMouseUpForSelect: jest.fn(),
  handleMouseLeaveForSelect: jest.fn()
}))


beforeEach(() => {
    
  useStore.setState({ 
    tables: {
      "1": {
        "name": "MockFirstTable",
        "attributes": {
          "13": {
            "name": "mockFirstAttribute",
            "type": "string",
            
          },
          "14": {
            "name": "mockSecondAttribute",
            "type": "string",
          }
        },
        "superclassId": ""
      },
    },
    resetStore: jest.fn(),
    uploadStore: jest.fn(),
    addAttribute: jest.fn(),
    removeAttribute: jest.fn(),
    removeTable: jest.fn(),
    addTable: jest.fn(),
    onAttributeTypeChange: jest.fn(),
    onAttributeNameChange: jest.fn(),
  })
    
  renderReadyComponent = (
    <AttributeTypeOptions tableId="1" attributeId="13" />
  );
});

describe('<AttributeTypeOptions />', () => {


  it('renders a drop-down list for each attribute', () => {

    render(renderReadyComponent );
    const { result } = renderHook(() => useStore());

    const stringType = screen.getAllByText(/string/i)[1]
    
    fireEvent.change(stringType.parentNode as Element, screen.getByText(/integer/i))

    expect(result.current.onAttributeTypeChange).toHaveBeenCalledTimes(1);
  });

  it('renders all types', () => {

    render(renderReadyComponent );
    const selectEl = screen.getByTitle(/Select attribute type/);

    ['primary_key', 'string', 'text', 'integer', 'float', 'decimal', 'datetime',
    'time', 'date', 'binary', 'boolean'].forEach((item) => {
      expect(selectEl.innerHTML).toContain(item)
    })
    
  });

  it('calls the handleMouseUpForSelect function', () => {

    render( renderReadyComponent );

    const buttonEl = screen.getAllByText(/^string$/)[0];

    fireEvent.mouseUp( buttonEl );
    
    expect(handleMouseUpForSelect).toHaveBeenCalledTimes(1)

  });
  
  it('calls the handleMouseLeaveForSelect function', () => {

    render( renderReadyComponent );

    const selectEl = screen.getByTitle(/Select attribute type/);
    
    fireEvent.mouseLeave( selectEl );

    expect(handleMouseLeaveForSelect).toHaveBeenCalledTimes(1)

  });
  it('renders correctly', () => {
    const renderedContainer = render(renderReadyComponent );
    expect(renderedContainer).toMatchSnapshot();
  });

});