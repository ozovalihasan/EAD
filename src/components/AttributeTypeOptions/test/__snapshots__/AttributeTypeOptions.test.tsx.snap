// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AttributeTypeOptions /> renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <select
        class="w-full"
      >
        <option
          value="primary_key"
        >
          primary_key
        </option>
        <option
          value="string"
        >
          string
        </option>
        <option
          value="text"
        >
          text
        </option>
        <option
          value="integer"
        >
          integer
        </option>
        <option
          value="float"
        >
          float
        </option>
        <option
          value="decimal"
        >
          decimal
        </option>
        <option
          value="datetime"
        >
          datetime
        </option>
        <option
          value="timestamp"
        >
          timestamp
        </option>
        <option
          value="time"
        >
          time
        </option>
        <option
          value="date"
        >
          date
        </option>
        <option
          value="binary"
        >
          binary
        </option>
        <option
          value="boolean"
        >
          boolean
        </option>
        <option
          value="references"
        >
          references
        </option>
      </select>
    </div>
  </body>,
  "container": <div>
    <select
      class="w-full"
    >
      <option
        value="primary_key"
      >
        primary_key
      </option>
      <option
        value="string"
      >
        string
      </option>
      <option
        value="text"
      >
        text
      </option>
      <option
        value="integer"
      >
        integer
      </option>
      <option
        value="float"
      >
        float
      </option>
      <option
        value="decimal"
      >
        decimal
      </option>
      <option
        value="datetime"
      >
        datetime
      </option>
      <option
        value="timestamp"
      >
        timestamp
      </option>
      <option
        value="time"
      >
        time
      </option>
      <option
        value="date"
      >
        date
      </option>
      <option
        value="binary"
      >
        binary
      </option>
      <option
        value="boolean"
      >
        boolean
      </option>
      <option
        value="references"
      >
        references
      </option>
    </select>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
